File: src/Api/Startup.fs
namespace Api
open System
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open System
open System.Collections.Generic
open System.IO
open System.Linq
open System.Threading.Tasks
open Microsoft.AspNetCore
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.AspNetCore.HttpsPolicy
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Microsoft.AspNetCore.Authentication.JwtBearer
open Microsoft.IdentityModel.Tokens
open System.Text

type Startup(config: IConfiguration) =

    member _.ConfigureServices(services: IServiceCollection) =
        let sectionSettings = config.GetSection("AppSettings")
        services.Configure<AppSettings>(sectionSettings) |> ignore
        let appSettings = sectionSettings.Get<AppSettings>()

        services.AddSingleton(appSettings) |> ignore
        services
            .AddAuthentication()
            .AddJwtBearer(fun options ->
                options.RequireHttpsMetadata <- false
                options.SaveToken <- false
                options.TokenValidationParameters <- TokenValidationParameters(
                    IssuerSigningKey = SymmetricSecurityKey(Encoding.ASCII.GetBytes(appSettings.Auth.Secret)),
                    ValidateIssuerSigningKey = true,
                    ValidIssuer = appSettings.Auth.Issuer,
                    ValidateAudience = false,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero)) |> ignore

        services.AddControllers() |> ignore

    member _.Configure (app: IApplicationBuilder, env: IWebHostEnvironment) =
        if (env.IsDevelopment()) then
            app.UseDeveloperExceptionPage() |> ignore

        app
            .UseAuthentication()
            .UseAuthorization() |> ignore

        app
            .UseHttpsRedirection()
            .UseRouting()
            .UseEndpoints(fun endpoints ->
                endpoints.MapControllers() |> ignore) |> ignore

File: src/Api/Startup.cs
namespace Api;
using System;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.IdentityModel.Tokens;
using System.Text;
using Microsoft.OpenApi.Models;


public class Startup(IConfiguration config)
{

    public void ConfigureServices(IServiceCollection services)
    {
        var sectionSettings = config.GetSection("AppSettings");
        services.Configure<AppSettings>(sectionSettings);
        var appSettings = sectionSettings.Get<AppSettings>()!;

        services.AddSingleton(appSettings);

        services.AddAuthentication()
            .AddJwtBearer(options =>
            {
                options.RequireHttpsMetadata = false;
                options.SaveToken = false;
                options.TokenValidationParameters = new TokenValidationParameters
                {
                    IssuerSigningKey = new SymmetricSecurityKey(Encoding.ASCII.GetBytes(appSettings.Auth.Secret)),
                    ValidateIssuerSigningKey = true,
                    ValidIssuer = appSettings.Auth.Issuer,
                    ValidateAudience = false,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero
                };
            });

        services
            .AddCors(options => options.AddDefaultPolicy(builder => builder.AllowAnyOrigin().AllowAnyHeader().AllowAnyMethod()));

        services
            .AddSwaggerGen(c => c.SwaggerDoc("v1", new OpenApiInfo { Title = "Api", Version = "v1.0" }))
            .AddControllers();
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
            app.UseSwagger().UseSwaggerUI(builder => builder.RoutePrefix = "swagger");
        }

        app.UseAuthentication().UseAuthorization();
        app.UseCors();

        app.UseHttpsRedirection()
            .UseRouting()
            .UseEndpoints(endpoints => endpoints.MapControllers());
    }
}